### PR TITLE
generate config files for atomspace

### PR DESCRIPTION
**Pull request by noskill**
My intention here is to collect and populate a list of all built atomspace libraries. This list is used to set necessary variables at find_package call.  
issue https://github.com/opencog/cogutil/issues/113

### CODE REVIEW

**Comment by noskill** [comment]
I couldn't figure out how to get list of targets from export set(AtomSpaceTargets), so i had to manually set another variable - atomspace_libs. Advice on how to use export set instead would be welcome.

**Comment by linas** [comment]
What problem is this solving? Why is this needed?

**Comment by noskill** [comment]
@linas Vitaly's idea in the issue https://github.com/opencog/cogutil/issues/113 is to replace Find* cmake files for opencog projects with config-file packages. This allows for other projects that use atomspace to find it without using FindAtomSpace.cmake, just by find_package(atomspace CONFIG) call.

**Comment by linas** [comment]
Ah. OK. I don't quite understand the mechanism. This seems to want some file `AtomSpaceTargets.cmake` but I don't see that file anywhere... is it autogened?  

For the pull request message, it would be a lot more useful if you wrote "replace Find* cmake files for opencog projects with config-file packages. This allows for other projects that use atomspace to find it without using FindAtomSpace.cmake, just by find_package(atomspace CONFIG) call."  

It explains what the pull request is for.  The current description  "My intention here ..." just leaves one to scratch one's head.

**Comment by noskill** [comment]
Yes, AtomSpaceTargets.cmake is autogenerated file with lines:
```
add_library(atomproto SHARED IMPORTED)
add_library(atombase SHARED IMPORTED)
etc..
```

AtomSpaceTargets.cmake also includes AtomSpaceTargets-<release|debug>.cmake  - another autogenerated file which sets properties:

```
# Import target "atomspace" for configuration "Release"
set_property(TARGET atomspace APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(atomspace PROPERTIES
  IMPORTED_LINK_INTERFACE_LIBRARIES_RELEASE "pthread;atomcore;truthvalue;/tmp/test/lib/libcogutil.so;/usr/lib/x86_64-linux-gnu/libboost_thread.so"
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/opencog/libatomspace.so"
  IMPORTED_SONAME_RELEASE "libatomspace.so"
  )
```

**Comment by noskill** [comment]
There is another option: after some more searching and cmake upgrade i found that this code works:
```
get_property(dirprops DIRECTORY "opencog/persist/guile/" PROPERTY BUILDSYSTEM_TARGETS)
```
also it is possible to recursively list directories and to collect all the targets which are going to be built.

So i want to collect all targets automatically with FILE(GLOB_RECURSE.. and GET_PROPERTY(..,  and leave hardcoded list as a fallback for old cmake. 
BUILDSYSTEM_TARGETS property supported since cmake 3.7. I suppose it's better than hardcoding  since we'll get correct list of libraries automatically.

**Comment by linas** [comment]
I think I would prefer the single, hardcoded list for two reasons. Vitaly already named one of them: single location, fewer worries about adding a cryptic line. ... because `set(blah blah blah)` really is cryptic -- its a kind of magic incantation, and if you don't already know about this pull request, and what it does, it just seems like magic mumbo-jumbo in the CMakefile.

Which leads to my other point: a simple list is easier to understand and easier to debug. You can look at a single list, and immediately recognize it as a list of libraries. You look at it, and you know what it is.  However, if you look at a bunch of lines like `FILE(GLOB_RECURSE..` you scratch your head and wonder - "what does this do? Why? Is it needed? What does it output? Why is it not doing what I think it should be doing? Line number 42 looks like a bug... let me change it" This kind of thinking makes future debugging more complicated, more time-consuming, more error-prone. 

The law of code is "there will be bugs", so the simpler we can keep things, the fewer bugs, and the less time lost trying to find bugs.  OpenCog is already complicated and hard to understand; I'm nervous about anything that makes it even more complicated. Try to find the simplest possible solution; and, before that, try to make sure that the problem that you are solving actually needs to be solved.  Not everything that seems like "a problem" actually is a problem.

File: lib/AtomSpaceConfig.cmake.in [hunk 1]
```diff
@@ -2,13 +2,63 @@
...
+set(ATOMSPACE_truthvalue_LIBRARY truthvalue)
+set(ATOMSPACE_atomutils_LIBRARY atomutils)
+set(ATOMSPACE_clearbox_LIBRARY clearbox)
+set(ATOMSPACE_lambda_LIBRARY lambda)
+set(ATOMSPACE_persist_LIBRARY persist)
+set(ATOMSPACE_query_LIBRARY query)
+set(ATOMSPACE_execution_LIBRARY execution)
+set(ATOMSPACE_ruleengine_LIBRARY ruleengine)
+set(ATOMSPACE_smob_LIBRARY smob)
+set(ATOMSPACE_unify_LIBRARY unify)
```
Comment (vsbogd) [inline lib/AtomSpaceConfig.cmake.in]: Following libraries are absent:
ATOMSPACE_LIBRARY - required by opencog/opencog
ATOMSPACE_PythonEval_LIBRARY - requried by opencog/benchmark
ATOMSPACE_persist-zmq_LIBRARY

File: lib/AtomSpaceConfig.cmake.in [hunk 2]
```diff
@@ -2,13 +2,63 @@
...
+    set(ATOMSPACE_persist-sql_LIBRARY persist-sql)
+    SET(ATOMSPACE_LIBRARIES
+        ${ATOMSPACE_LIBRARIES}
+        ${ATOMSPACE_persist-sql_LIBRARY}
+    )
+    SET(HAVE_PERSIST_SQL 1)
+    ADD_DEFINITIONS(-DHAVE_PERSIST_SQL)
+ENDIF (persist-sql)
+
+# persist-zmq is optional
```
Comment (vsbogd) [inline lib/AtomSpaceConfig.cmake.in]: incorrect comment, should be ```# zmqatoms is optional```

File: lib/AtomSpaceConfig.cmake.in (hunk 1 above)
Comment (noskill) [reply lib/AtomSpaceConfig.cmake.in]: It seems there is no persist-zmq library, where it is defined? opencog/persist/zmq/atomspace/CMakeLists.txt defines only  zmqatoms..

File: lib/AtomSpaceConfig.cmake.in (hunk 2 above)
Comment (noskill) [reply lib/AtomSpaceConfig.cmake.in]: Thanks! 
I updated comments

File: lib/AtomSpaceConfig.cmake.in (hunk 1 above)
Comment (vsbogd) [reply lib/AtomSpaceConfig.cmake.in]: Yes, I have double checked, don't see it as well.

**Comment by vsbogd** [comment]
@linas , @ngeiswei, I think it is ready to be merged, if you don't have any objections I would merge it.

**Comment by ngeiswei** [comment]
I didn't look at it but I trust your judgment, @vsbogd. Just wait for @linas' approval.

**Comment by linas** [comment]
I have no objections.

**Comment by vsbogd** [comment]
Thanks!

**Pull request merged**